train_gpt2.cu

```diff
@@ -1572,6 +1607,42 @@ void matmul_backward(floatX* dinp, floatX* dweight, floatX* dbias,
     NVTX_RANGE_FN();
     float one = 1.0f;
     float zero = 0.0f;
+
+    // backward to bias, if given, does a +=
+    if (dbias != NULL) {
+        // Each warp is responsible for 32 * "x128::size" = 256 OCs at BF16 (OC must be a multiple of 256!)
+        // Block size is 512 threads (16 warps) and we reduce those 16 values into 1 at the end
+        // blockDim.x is 32 --> single warp being responsible for those 256 OCs
+        // blockDim.y is 16 --> 16 parallel independent warps processing the same OCs for different BTs
+        // gridDim.x is OC / 256 --> each block processes 256 OCs
+        // grimDim.y is max(1, (cuda_num_SMs * cuda_threads_per_SM) / (512 * gridDim.x)); --> fill up the entire GPU!
+        const int warp_size = 32;
+        const int block_size = 512;
```

Comment:
Was testing out on the A100's making this 1024 would make a substantive difference, made a CR but didnt realize you had one out already

Reply:
1024 might be a terrible choice on non-A100/H100 cards (i.e., all those with 1.5k instead of 2k threads). Not sure if we should hard-code this. Haven't benchmarked this praticular kernel, but a slowdown of 33% wouldn't be unexpected.

Reply:
@ChrisDryden , you're right 1024 was a big win for the previous version of this kernel, but my new kernel6 doesn't seem to benefit, have you tried it on your end?

@ngc92 we could query "max_warps_per_multiprocessor" to determine that, for kernels that benefit from the largest possible block size at maximum occupancy, 768 would be optimal on non-A100/H100 cards (except Turing which is limited to 32 rather than 48!) - where it gets complicated is that depending on L2 cache size, I think some kernels like fused_classifier might actually benefit from *lower* occupancy... maybe something for me to investigate when I eventually start working on my "automatic sweeping of L2 cache residency per-kernel" idea.